File: src/core/account.ts
export enum AccountType {
  Asset = 'Asset',
  Liability = 'Liability',
  Equity = 'Equity',
  Income = 'Income',
  Expense = 'Expense',
}

export class Account {
  constructor(
    public readonly code: string,
    public readonly name: string,
    public readonly type: AccountType,
    public readonly normalBalance: 'debit' | 'credit' = type === AccountType.Asset || type === AccountType.Expense ? 'debit' : 'credit'
  ) {}

  toString(): string {
    return `${this.code} ${this.name} (${this.type})`;
  }

  /** Serialize for persistence / roundtrip (pairs with fromJSON). */
  toJSON(): { code: string; name: string; type: AccountType } {
    return { code: this.code, name: this.name, type: this.type };
  }

  /** Reconstruct from serialized form. Derives normalBalance from type. */
  static fromJSON(j: any): Account {
    if (!j || typeof j !== 'object' || !j.code || !j.name || !j.type) {
      throw new Error('Account.fromJSON: invalid account shape');
    }
    return new Account(j.code, j.name, j.type as AccountType);
  }
}

File: reference-implementations/python/ledger/account.py
"""
Python canonical: Account + AccountType. Mirrors src/core/account.ts exactly.
"""

from enum import Enum
from typing import Any


class AccountType(str, Enum):
    Asset = "Asset"
    Liability = "Liability"
    Equity = "Equity"
    Income = "Income"
    Expense = "Expense"


class Account:
    def __init__(self, code: str, name: str, type_: AccountType, normal_balance: str = None):
        self.code = code
        self.name = name
        self.type = type_ if isinstance(type_, AccountType) else AccountType(type_)
        if normal_balance is None:
            self.normal_balance = "debit" if self.type in (AccountType.Asset, AccountType.Expense) else "credit"
        else:
            self.normal_balance = normal_balance

    def __repr__(self) -> str:
        return f"{self.code} {self.name} ({self.type.value})"

    def to_json(self) -> dict:
        return {"code": self.code, "name": self.name, "type": self.type.value}

    @classmethod
    def from_json(cls, j: Any) -> "Account":
        if not isinstance(j, dict) or not j.get("code") or not j.get("name") or not j.get("type"):
            raise ValueError("Account.from_json: invalid account shape")
        return cls(j["code"], j["name"], AccountType(j["type"]))
